### ISSUE TITLE
port of tf.image.ssim as DSSIM to pure keras

### ISSUE THREAD

**Issue by iperov** [keras-team/keras-contrib#464]
I made this port for my own project in order to work on plaidML keras backend.

Tensorflow's ssim is better than current keras-contrib's DSSIMObjective. 

feel free to adapt it to keras-contrib project 

(updated)

```
class DSSIMObjective:
    """Computes DSSIM index between img1 and img2.
    This function is based on the standard SSIM implementation from:
    Wang, Z., Bovik, A. C., Sheikh, H. R., & Simoncelli, E. P. (2004).
    """

    def __init__(self, k1=0.01, k2=0.03, max_value=1.0):
        self.__name__ = 'DSSIMObjective'
        self.k1 = k1
        self.k2 = k2
        self.max_value = max_value
        self.backend = K.backend()

    def __int_shape(self, x):
        return K.int_shape(x) if self.backend == 'tensorflow' else K.shape(x)

    def __call__(self, y_true, y_pred):
        ch = K.shape(y_pred)[-1]

        def _fspecial_gauss(size, sigma):
            #Function to mimic the 'fspecial' gaussian MATLAB function.
            coords = np.arange(0, size, dtype=K.floatx())
            coords -= (size - 1 ) / 2.0
            g = coords**2
            g *= ( -0.5 / (sigma**2) )
            g = np.reshape (g, (1,-1)) + np.reshape(g, (-1,1) )
            g = K.constant ( np.reshape (g, (1,-1)) )
            g = K.softmax(g)
            g = K.reshape (g, (size, size, 1, 1)) 
            g = K.tile (g, (1,1,ch,1))
            return g
                  
        kernel = _fspecial_gauss(11,1.5)

        def reducer(x):
            return K.depthwise_conv2d(x, kernel, strides=(1, 1), padding='valid')

        c1 = (self.k1 * self.max_value) ** 2
        c2 = (self.k2 * self.max_value) ** 2
        
        mean0 = reducer(y_true)
        mean1 = reducer(y_pred)
        num0 = mean0 * mean1 * 2.0
        den0 = K.square(mean0) + K.square(mean1)
        luminance = (num0 + c1) / (den0 + c1)
        
        num1 = reducer(y_true * y_pred) * 2.0
        den1 = reducer(K.square(y_true) + K.square(y_pred))
        c2 *= 1.0 #compensation factor
        cs = (num1 - num0 + c2) / (den1 - den0 + c2)

        ssim_val = K.mean(luminance * cs, axis=(-3, -2) )
        return K.mean( (1.0 - ssim_val ) / 2.0 )
```

**Comment by gabrieldemarmiesse** [comment]
Thanks! A pull request is welcome if you think you can add the documentation and the tests.

**Comment by iperov** [comment]
I thought someone will adapt it ...
it has no channels check on compile time, and it does not work with variable channel dims ...

variable dims requires kernel
`g = np.tile (g, (1,1,ch,1))` also be variable

**Comment by iperov** [comment]
updated main post

**Comment by gabrieldemarmiesse** [comment]
Maybe you can use `K.softmax` to simplify your code even further? What do you think?

**Comment by iperov** [comment]
done

**Comment by iperov** [comment]
upd, anyway it should be tested by ppl :)

**Comment by iperov** [comment]
but it works fine on tensorflow and plaidml, 
using it in my DeepFaceLab
